### ISSUE TITLE
SoftIEC Command cd:<- / cd<-

### ISSUE THREAD

**Issue by radius75** [GideonZ/1541ultimate#392]
Referring to the discussion on Messenger.

The SoftIEC command **cd<-** or **cd:<-** takes us to the directory defined by the Default Path option.

This seems to be consistently associated with the **cd<-** command of CMD storage devices. Where it takes us to the root of the partition.

However, the **cd:<-** command in CMD takes us to a folder called **<-**

This creates an exception.

If it is not planned to maintain compliance with the standard previously introduced by CMD, please ignore the report.

I know that there are already ready-made applications that use the command
**cd:<-** in SoftIEC
Which complicates the introduction of some standardization here. This would involve updating this software.

This is a non-critical issue. It does not affect the operation of other programs. It is merely confusing for the user who uses CMD and SoftIEC commands at the same time.

**Comment by markusC64** [comment]
> This seems to be consistently associated with the cd<- command of CMD storage devices. Where it takes us to the root of the partition.

No, it takes you one director level up. That is, it changes to the parent directory. Both have checked the manual and tested.

> However, the cd:<- command in CMD takes us to a folder called <-
Well, on SD2IEC, »cd:←«  is the same as »cd←«. Therefore I think best is to avoid »←« as directory name.. Note we cannot be compatible to both without a configuration setting.  And since CMD has not a home or default directory or whatever you call it, we need a syntax to switch to it. Perhaps UNIX like the empty path (I doubt CMD has used the empty path, so that could work).

> If it is not planned to maintain compliance with the standard previously introduced by CMD, please ignore the report.

We do want to make it compatible, don't we? For a few things it makes sense to make them configurable.

**Comment by markusC64** [comment]
> Which complicates the introduction of some standardization here. 

If you consider classic C64 hardware (which includes 1541 speeders), you'll note you cannot be compatible to all of them. I think and that is only my opinion, the best to do is providing a possibiliy to write software that works on all of them. A programmer can always chosse a non compatible command, but there should be no need for him.

**Comment by markusC64** [comment]
In order not to forget it: You change to the root of the partition on CMD devices with »cd//«.

**Comment by radius75** [comment]
All clear. My mistake. I built a directory structure only for 1 lv. down. Hence, I wrongly assumed that cd< always moves to root.

**Comment by GideonZ** [comment]
Yup... we need to differentiate between CD and CP. Does CP also allow paths?

**Comment by markusC64** [comment]
No, "cp" is "change partition" and only accepts a number as parameter. Nothing else. And if you write "cP" (uppercase P), the number is not written in PETSCII, but as binary value. Except that technical difference, "cP" is the same as "cp".
CMD partitions are in the sense like "gparted" or "diskpart" or "fdisk"... and there is really a C64 tool tthat creates them.

**Comment by GideonZ** [comment]
Okay.. Then how do we define what are the roots of a partition, if the CP
command doesn't support a path? How do you specify (programmatically) where
the partitions are located in the file system?

**Comment by markusC64** [comment]
Since CMD and IDE64 don't have that concept,  I would say: Via file browser of the ultimate. You know I have already started changes which certainly need refactoring. And pretty printing. Like "Set dir. here" already present in F5, ... SoftIEC..., I have added "Set root here". Both open a window where the user enters the partition number to change.

And for partitions 1 to 12, he can also choose "Cfg. root here", which copy the path to the SoftIEC settings. So it is available at poweron.
Note we have asked some more or less random users: Most do just use 1 partitions. Other think that 4 to 8 partitions are enough. "Device Manager" needs partition 11 (with "/" as the root and some working directory as default directory). So 12 permanent partitions are probably enough.

Note that this is only one possibility - but one that makes it simple for the user. We can decide to use another user interface if we have a better idea. But it should be a function of the configuration area together with the file manager (to remember and to type in paths are is not very convenient, is it?)

We could also add a command that is not used by any pf the speeders and CMD, IDE64 and SD2IEC. Perhaps "cr" like change root. So that the C64 code can set them.

**Comment by markusC64** [comment]
@radius75 If @GideonZ thinks it will help, I would start soon writing down all commands we should support in an Excel sheet. With comments like path allowed, path necessary or path forbidden. The documentation is quite quiet about where wildcards are supported and if they're supported, whether they affect all files or only the first file found.
If you want to help, you can then help to find that out.

**Comment by radius75** [comment]
Additional option "set default path as root" set to off by default?
So that it is impossible to go cd// higher than the user wants.
Is this any solution?

**Comment by markusC64** [comment]
That is to be discussed. In my, well I might call it Proof Of Concept, you can combine virtual root (I often call it chroot after the Unix command) and default path.

/path/to/foo/
is a default path with virtual root "/"

/path/to/root//
is a default path which equals the virtual root.

/path/to/root//and/within/some/path/
has virtual root »/path/to/root/« and within it »/and/within/some/path/« as a default path. So you just need to adapt the extra slash to change that.
Of cause, a UI might give the user the few possibilities to choose from...  So that he cannot do it wrong when using the file browser.

Note that this way it does not need really more storage than not having the arbitraty virtual root. So that is space optimized storage of configuration.


Dunno what you and Gideon think, but I like the extra flexibility that this provides.

**Comment by radius75** [comment]
What about something like an index file?
In which you should write the path to each partition. Path to the folder or inside .dnp.
255 records could be saved in such a file, to address partitions 1 to 255 in turn.
255 is the max for CMDHD

**Comment by GideonZ** [comment]
Yes, this is exactly how I discussed it before with Mario. But never got
the change to implement it.
Making such a file is easy, but making it also user-friendly editable in
the menu is a lot harder.

**Comment by radius75** [comment]
Menu F5 and 'set partition no. XXX here'?
With saving changes to the appropriate record in the index file.

**Comment by markusC64** [comment]
Well, I see two problems with an index file. If we manage to solve them, we can use it, too.

1) What if two index files on two USB sticks have conflicting index files and you insert both of them?
2) How to create a Partition with root "/Temp"?

Not really a problem to solve:
3) Make sure it works with "/Flash", too. Current device manager makes a »cd:/Flash" and tries to load a file. If this fails, it tries the same from the default directory. Perhaps Bart would like to change it to real partitions with a later release, Or others will do something like that. Better to support that.

**Comment by radius75** [comment]
Solution proposal for 1.
Since there is only one SoftIEC device.

Multiple index files can be stored to run different drives with different sets of partitions. Under different names.
e.g. drive1.sys drive2.sys.... or something. (.idx is already provided for TAP files)

But of course not loaded at the same time.

The path to the currently selected index file is manually set in the options.
Then this would be the only path needed to enter in the options. All other necessary paths to all selected partitions would be read from the index file.

**Comment by radius75** [comment]
Basic question
Is there anything stopping the selected partitions (saved in idx) from being in different folders, or even on different USB sticks? Some on USB0 and some on USB1?

**Comment by markusC64** [comment]
I assume a text file with at least one path per partition. Extra info in that file as needed.

Then it depends. If the configuration file if defined to contain paths relative to the root of the usb stick, then you cannot leave the usb stick.
If the configuration file if defined to contain absolute paths, then you must plug in the usb stick in the right usb connector. If you use the wrong usb connector, e. g. /Usb0 changes to /Usb1.

Both have pros and cons.

**Comment by markusC64** [comment]
One more thing to note: If we don't want to break device manager without a possibility to repar it by asking Bart tto patch device manager: We need the partition configuration being active at power on time. Device manager is an emulated cartridge for the C128 that starts software at power on time.

**Comment by markusC64** [comment]
Well, I think I have to think about that.

I still prefer having the partion table on the flash. But it might be possible to place that file just into the flash file system as an option. So when done properly, that is no show stopper.

You see, autoload partition tables is also something I have to think about it. Curently I am worried that the following situation might get a problem: I need one file from a different usb stick. So I insert it in order to copy the file from it to let's say /Temp. Well, in that case I would not like that it disturbs my partitions.
Partition file always on "/Usb0" resp. "/Flash" would avoid that clearly. 


BTW: It's natural to have partitions spread on multiple USB sticks. CMD HD does something similar: It has a SCSI bus where multiple hard disks and similar (like ZIP disk) can be attached. So partitions are spread over multiple data storing devices.

**Comment by GideonZ** [comment]
Hmm... I think we misunderstood each other. In particular, I think it does *not* mess up your partitions.
Take for example this config file on /Flash
```
{  "partitions": [
    { "number": 11,
       "path": "/Temp"
    } ]
}
```
This would specify the temp partition on 11. Then, when the USB stick is recognized in port 0, it loads the config file on the root of the USB, which contains:
```
{  "partitions": [
    { "number": 1,
       "path": "SomeDir"
    },
    { "number": 2,
       "path": "SomeOtherDir"
    } ]
}
```
Then, your partition table would look like this:
[ 1: "/Usb0/SomeDir",
  2: "/Usb0/SomeOtherDir",
11: "/Temp" ]

When you remove the USB stick, the list of partitions is again updated to:
[11: "/Temp"]

Like this you can have multiple USB sticks, with multiple sets of partitions, which are automatically merged.

**Comment by markusC64** [comment]
I see. The point is something different. Is a running C64 program necessarily prepared for partitions disappering?

And is mount time reproducible? If /Usb0 and /Usb1 are connected at poweron, the order seems to be quite random (but not necessarily equally distributed). Having random partitions - well, not the best idea. Would have to avoid that situation.

**Comment by GideonZ** [comment]
It doesn't matter. If you have conflicting configurations, you have conflicting configurations. It is not something you should try to solve by being smarter than organized.

**Comment by GideonZ** [comment]
> I see. The point is something different. Is a running C64 program necessarily prepared for partitions disappearing?

In the same way as programs are prepared for floppies being removed from a disk drive, I guess?

**Comment by GideonZ** [comment]
We could even add a priority field in the config file to make it even fancier... to know which one should take preference, if conflicts occur.

**Comment by GideonZ** [comment]
I think you are overcomplicating things. If you have conflicting configurations, you have conflicting configurations. Error out if you wish and that's it. No need for clumsy heuristics. Bail, show a message, so you can fix your issue in a way it should be fixed.

Same for Geos. Geos cannot prevent you from removing a floppy disk. Neither can the ultimate prevent you from removing a USB stick. Once the Usb is removed the files on it cannot be accessed. So this means the files in the removed partitions cannot be accessed.

**Comment by markusC64** [comment]
Probably yes. Perhaps we should just give it a try, Knowing we can improve it at any time. E. g. if it is found that auto adding partitions should be restricted, we can restrict it.

**Comment by radius75** [comment]
Coming back to "cd<-" for a moment

I looked at how it works on Vice in CMDHD emulation
I don't know for sure if we can trust that this emulation behaves like a real device in every case.

**cd<-** -directory lv up
**cd:<-** -directory lv up (!)
**cd:<-name** -go to directory named **<-name**

This means that a directory with only **<-** in its name cannot be accessed with the **cd:<-** command. (But you can enter in another way)

And the CMD manual doesn't mention anything about it.
It doesn't even give an example of using the command this way with a colon.

![obraz](https://github.com/GideonZ/1541ultimate/assets/1929111/75ab6f03-ea5a-4461-9814-0ad7ad7fa371)

**Comment by markusC64** [comment]
> I looked at how it works on Vice in CMDHD emulation
> I don't know for sure if we can trust that this emulation behaves like a real device in every case.

Well, for that case it's just that the emulation executes original CMD HD Dos. That cannot really go wrong. It's different when we run into issues which look like timing - in that case we cannot trust an emulation too much. The same might be true when we try to attach more than one device to the internal SCSI bus of the CMD HD (yeah, that is possible, and documentation on that is really rare). 
Anyway, if there is some doubt, I know a few having original CMD HD and CMD Ramlink. So there's an option when needed.

**Comment by markusC64** [comment]
A first idea what commands to support: https://1drv.ms/x/s!AkgcxWekAgAJhnVdEL1IsP--gMmB?e=PKlaNM

**Comment by markusC64** [comment]
No comments on my spreadsheet? There cannot be all well, because some information is missing in the spreahsheet.

**Comment by GideonZ** [comment]
Markus, I am sorry that I have gotten so busy with other things that I had totally forgotten about this. I will have a look at the spreadsheet shortly.

**Comment by markusC64** [comment]
No problem, I deal that like Terence: "Homo sum, humani nil a me alienum puto" (I am a human being, nothing human, I think, is alien to me).

If you need write access to the spreadsheet and you also have a microsoft account, I can sent you a writable share.

**Comment by bvl1999** [comment]
> One more thing to note: If we don't want to break device manager without a possibility to repar it by asking Bart tto patch device manager: We need the partition configuration being active at power on time. Device manager is an emulated cartridge for the C128 that starts software at power on time.

Currently, it already waits for a little while after power-on to give the UII+ time to settle (scan usb storage, etc). This is one reason why it does this 'quick ram test' after power-on.

Regardless, changing DM to support better compatibility with CMD HDD commands is in itself not a big problem, but the way it uses partitions currently is rather peculiar, and primarily done to avoid any interference with the current working directory as used by most software, and to ensure DM can find its stuff without having to know much about your setup. 

That is of course solvable, without making any use of partitions at all, but that would require at least one of 2 things:
- being able to query the current working directory, so DM can save and later restore it. The 'restore' is *REALLY* important because DM can suspend software, do save current state to /temp, and later resume it. That is in turn needed to allow changing disk images in 80 column mode, without needing telnet or any such thing.

**Comment by markusC64** [comment]
Another idea can be borrowed from the SD2IEC: For a special partition (its eeprom partition) you can use "!" instead of a partition number.
We could do something similar: Partition "!" could give access to the raw file system, that is with root "/".

**Comment by bvl1999** [comment]
> You know this is not portable to other devices. But I agree, that is a nice extra feature.

Thing is, I currently 'abuse' the fact every 'partition' can access all of the filesystem tree, and has its own 'current directory' to work around not being able to query and restore the current workind directory.

DM cannot provide the functionality it does unless such a workaround, or the ability to query the current path exists, so it is more than a 'nice to have' for me, it is required if it is no longer possible to access the full filesystem from within every partition. 

Don't get me wrong, I'm totally prepared to make the required changes to DM to work with this, but, only if a solution for this is provided, else it is simply a no-go for me.

**Comment by bvl1999** [comment]
> > if it is no longer possible to access the full filesystem from within every partition.
> 
> From every partiton? Really? Are you dynamically using partiion numbers?

No, I have picked partition 11 to use for this, but this means I do use the fact you can currently access the entire filesystem from more than one partition. (11 was based on people probably never using such a high partition number themselves, but the actual number doesn't really matter much)

What I did say is that I need a way to temporary change the working directory to any place in the filesystem tree (so, including /temp, /flash, and any location on usb storage where people might store disk images), without changing the working directory the user cd'ed into, and am currently abusing the way 'partitions' work for that. That works well, but it is not really a very nice solution, I'd call it a hack, which is why I'd be happy to change this if there is a better alternative.

> 
> Anyway, I like the idea of getting the current dorectory. I will create an entry in the spreadsheet...

Thanks, that would be helpfull for me, and I'm sure also for others.

Btw, Xander's dm boot menu will also be affected by this, will need to do an 'impact analysis' to get all the details.

**Comment by markusC64** [comment]
Well, then requiring the user to configure partition 11 properly (that is to the root where the devices are) is a solution. Note that I am not saying it is the best solution.

But we're free to think about a better solution.

**Comment by bvl1999** [comment]
> Well, then requiring the user to configure partition 11 properly (that is to the root where the devices are) is a solution. Note that I am not saying it is the best solution.
> 
> But we're free to think about a better solution.

The more config options, especially those which change the behavior of details of soft iec behavior, the more complex things become for end-users, and for people who support those end-users. 

I'd strongly prefer consistent and predictable behavior, without bothering end-users with even more config options which make a lot of sense to us maybe, but not to the average user.

**Comment by GideonZ** [comment]
Agreed, we should eliminate all configuration items in relation to behavior. The only allowable configuration is a predefined set of partition mount points.

**Comment by markusC64** [comment]
I also agree. Since Barth and Xander are willing to adapt their software, we do not need a compatibility mode. Other software is quite rare - and if there is an interest, someone will adapt it.
But one or two configuration options might make sense - I post them here to discuss them:
(1) Enable Long File Name Support, that is File names longer than 16 characters. The iec device of vice offers them. But many software would produce a buffer overflow when getting those filenames, so one might want to turn them off and only on, when needed.
(2) Enable / Disable Jiffydos. Today I would like to have it. CBM Kernal, Jiffydos Kernal, PAL, NTSC, all seem to make a difference. Would be great if software can be tested with the same kernal but jiffydos enabled/disabled. BTW: Don't ask for to many details, they are unknown. We only have that observation.

**Comment by bvl1999** [comment]
> I also agree. Since Barth and Xander are willing to adapt their software, we do not need a compatibility mode. Other software is quite rare - and if there is an interest, someone will adapt it. But one or two configuration options might make sense - I post them here to discuss them: (1) Enable Long File Name Support, that is File names longer than 16 characters. The iec device of vice offers them. But many software would produce a buffer overflow when getting those filenames, so one might want to turn them off and only on, when needed. (2) Enable / Disable Jiffydos. Today I would like to have it. CBM Kernal, Jiffydos Kernal, PAL, NTSC, all seem to make a difference. Would be great if software can be tested with the same kernal but jiffydos enabled/disabled. BTW: Don't ask for to many details, they are unknown. We only have that observation.

I see the point of both those options. They are also not 'obscure' in the sense of talking about things which won't make sense to the average user.

There is really only a few other bits of software, one of which, drabrowse/copy contains updates by Xander and me to make it work with soft iec, but it already works with sd2iec, cmd hdd, and compatible, so that is a really trivival fix I'm sure either Xander or me could submit. 

But someone should get in contact with Martijn Wehrens at some point about this because of his CP/M U2 tools which talk to the softiec target on uci. I think Xander has been in contact with him relatively recently?

**Comment by xahmol** [comment]
Yes, I have been in contact with Martijn, he also discusses his CP/M work here (c128old is his nick)
https://c-128.freeforums.net/post/7508/thread

Can mail him and point him to this thread if you want.

**Comment by xahmol** [comment]
And yes,I myself would prefer a simple code change in my software over blocking the very worthy goal of getting to the standard commands without a myriad of config options.
As Bart indicates, the cd change is rather trivial to implement. But it indeed has to be implemented in all software that implements UCI based file browsing.

**Comment by mwehrens** [comment]
sorry for the long text..

From the discussion so far, I don't see many issues, because for my code UCI recognition is via the U2 command interface. SD2IEC is treated as a 'slow' device and CMDHD as a 'fast' device. The one path (sd2iec, uci), or the drive-partitionlinks (cmdhd) are set using dedicated cpm-commands.

I'm not using 'CP' actively for UCI. The implementation I have is for CMDHD to dynamically switch between (max 5) images on drives A-E. This happens to be compatible with SD2IEC partitions.

With regards to 'make UCI like CMDHD', and what SD2IEC is doing:
Partitions on SD2IEC allow to switch to (at least 5, in my test) partitions on 1 FAT32 SDcard. SD2IEC 'remembers' the path it was using per partition. If you have partition 1 with 'CD:cpm1.dnp' and partition 2 with 'CD:cpm2.dnp', then you can CP1 and CP2 between them staying in the .dnp files. But I don't see this being used.
Suppose one makes the Ultimate-UCI switch likewise (meaning, to 'remember' where one was per partition) then it would be possible to mimic the SD2IEC behavior and thus appear like the CMD-HD partitions (where you are always inside an image).
Can of worms is the set of $-P, G-P etc. interfacing commands, including partition-types (D81Cpm anyone?)

My 2euro-c to the "UCI wishlist':
While UCI will benefit from behaving (more) like CMDhd and or SD2IEC, there still is the IEC communication overhead (with the buffer etc). For instance, a CMDHD can read 16K in 1 burst-message to the drive on channel15 (and I use that in my CP/M implementation to max out the burst-IEC). So, for my part I rather would enhance Ultimate UCI to be able to send a 'Fast IEC command' with the nr. of sectors to read and the DMA memory where to put it/get it. That would make DNP files 'fly' on CPM (making them as fast as the REU).

NOTE: Ultimate will (today) happily execute an U1 command in (any) image without buffer (fine with me) but sd2iec will not.
Given c128, "monitor", uci on 12=$C, path in any image, When: '@c U1 0 0 1 0', Then: 0, OK 0 0
Given c128, "monitor", sd2iec on 9, path irrelevant, When '@a U1 0 0 1 0', Then: 70,NO CHANNEL,0,0

Before UCI, I was using the DOS target... I'm flexible for implementation.
Although my current implementation for UCI read of 1 block assumes behavior like: "U1 ..." then input 32, accept, input 224. The protocol with 'is there a byte' status check is nice and all but for now I'm ignoring the cmd-i/f-statemachine during a read. For reference, the IEC Burst i/f tests the CIA all the time, but I imagine Ultimate can assume more control here.

@markusC64 : can you add a column in the xls indicating what currently works on Ultimate?
How many of the commands are critical? For instance: a lot of sd2iec have no battery so do not implement "t-r" /"t-w".

**Comment by markusC64** [comment]
> NOTE: Ultimate will (today) happily execute an U1 command in (any) image without buffer (fine with me) but sd2iec will not.

Oh, we have to check where the data is stored. That is it is not overwriting random memory... Well, "U1" is also not working correctly. The sample reading the BAM from the 1541 test/demo disk is not working in the release firmware.

**Comment by bvl1999** [comment]
So a thorough read if the kernal rom shows a 'record' or command always ends on EOI, and EOI can be triggered from clrchn, ckout and close.

Also the only situations where it can be ignored are clrchn and ckout on sequential files, in that case data can be appended to the existing buffer. That works because you need to close a file or end up with a splat file.
